src/orange/detail/BIHMaker.hh

```diff
@@ -0,0 +1,116 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022-2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/detail/BIHMaker.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <vector>
+
+#include "orange/BoundingBox.hh"
+#include "orange/OrangeData.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Create a bounding interval hierarchy from supplied bounding boxes.
+ *
+ * This implementation matches the structure proposed in the original
+ * paper [1]. Partitioning is done on the basis of bounding box centers using
+ * the "longest dimension" heuristic. All leaf nodes contain either a single
+ * volume id, or multiple volume ids if the volumes have bounding boxes that
+ * share the same center.
+ *
+ * [1] C. Wachter, Carsten and A. Keller, "Instant Ray Tracing: The Bounding
+ * Interval Hierarchy" Eurographics Symposium on Rendering, 2006,
+ * doi:10.2312/EGWR/EGSR06/139-149}
+ */
+class BIHMaker
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using VecBBox = std::vector<BoundingBox>;
+    using Storage = Collection<LocalVolumeId,
+                               Ownership::value,
+                               MemSpace::host,
+                               OpaqueId<LocalVolumeId>>;
+    using VecNodes = std::vector<BIHNode>;
+    //!@}
+
+  public:
+    // Construct from vector of bounding boxes and storage for LocalVolumeIds
+    explicit CELER_FUNCTION BIHMaker(VecBBox bboxes, Storage* storage);
+
+    // Create BIH Nodes
+    CELER_FUNCTION VecNodes operator()();
+
+  private:
+    /// TYPES ///
+
+    using VecReal3 = std::vector<Real3>;
+    using VecIndices = std::vector<LocalVolumeId>;
+    using PairVecIndices = std::pair<VecIndices, VecIndices>;
+    using VecAxes = std::vector<Axis>;
+    using AxesCenters = std::vector<std::vector<double>>;
+
+    struct Partition
+    {
+        Axis axis = Axis::size_;
+        double location;
```

Comment:
We're storing some things as reals, some as doubles, and (in the final data) some as floats...

Reply:
I will make this more consistent, but note that this Partition struct is only used during construction. It stores only one value bisecting space, whereas with BIH a partition is both an upper and lower bound. My plan is to use real_types for intermediate stuff and store the final bounds as floats.